ansible_rulebook/websocket.py

```diff
@@ -12,126 +12,216 @@
-                "feedback websocket %s connection closed, will retry...",
-                websocket_address,
+                    await asyncio.sleep(int(backoff_delay))
+                # Increase delay with truncated exponential backoff.
+                backoff_delay = backoff_delay * BACKOFF_FACTOR
+                backoff_delay = min(backoff_delay, BACKOFF_MAX)
+                continue
+        else:
+            # Connection succeeded - reset backoff delay
+            backoff_delay = BACKOFF_MIN
+            refresh = True
+
+    return result
+
+
+async def request_workload(activation_id: str) -> StartupArgs:
+    return await _connect_websocket(
+        handler=_handle_request_workload,
+        retry=False,
+        activation_id=activation_id,
+    )
+
+
+async def _handle_request_workload(
+    websocket: WebSocketClientProtocol,
+    activation_id: str,
+) -> StartupArgs:
+    logger.info("workload websocket connected")
+    await websocket.send(
+        json.dumps(dict(type="Worker", activation_id=activation_id))
+    )
+
+    project_data_fh = None
+    response = StartupArgs()
+    while True:
+        msg = await websocket.recv()
+        data = json.loads(msg)
+        if data.get("type") == "EndOfResponse":
+            break
+        if data.get("type") == "ProjectData":
+            if not project_data_fh:
+                (
+                    project_data_fh,
+                    response.project_data_file,
+                ) = tempfile.mkstemp()
+
+            if data.get("data") and data.get("more"):
+                os.write(project_data_fh, base64.b64decode(data.get("data")))
+            if not data.get("data") and not data.get("more"):
+                os.close(project_data_fh)
+                logger.debug("wrote %s", response.project_data_file)
+        if data.get("type") == "Rulebook":
+            response.rulesets = rules_parser.parse_rule_sets(
+                yaml.safe_load(base64.b64decode(data.get("data")))
             )
-        except CancelledError:
-            logger.info("closing feedback websocket due to task cancelled")
-            return
-        except BaseException as err:
-            logger.error(
-                "feedback websocket error on %s err: %s", event, str(err)
+        if data.get("type") == "ExtraVars":
+            response.variables = yaml.safe_load(
+                base64.b64decode(data.get("data"))
             )
+        if data.get("type") == "ControllerInfo":
+            response.controller_url = data.get("url")
+            response.controller_token = data.get("token")
+            response.controller_ssl_verify = data.get("ssl_verify")
+    return response
+
+
+@dataclass
+class EventLogQueue:
+    queue: asyncio.Queue = field(default=None)
+    event: dict = field(default=None)
+
+
+async def send_event_log_to_websocket(event_log: asyncio.Queue):
+    logs = EventLogQueue()
+    logs.queue = event_log
+
+    return await _connect_websocket(
```

Comment:
@bzwei previously we were using [async for](https://github.com/ansible/ansible-rulebook/blob/c72a42087508e2aa03c0a183cf140dcc2b5aa093/ansible_rulebook/websocket.py#L94) to put the connection in a loop so when it got interrupted the for loop will create a new connection. In the new design how is the loop recreated.

Reply:
There is a `while True:` loop inside `_connect_websocket` function